{
 "metadata": {
  "name": "",
  "signature": "sha256:7bd9f3f5ac6af6115f8debab268d45dacb7713debc66a0363c214ce257084ee8"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from collections import defaultdict\n",
      "lowest_level_dict = defaultdict(list)\n",
      "with open('ensembl2reactome.txt') as f:\n",
      "    for line in f:\n",
      "        sep_line = line.strip().split('\\t')\n",
      "        if sep_line[5] == 'Homo sapiens':\n",
      "            lowest_level_dict[sep_line[1]].append(sep_line[0])\n",
      "        \n",
      "lowest_level_dict = dict(lowest_level_dict)    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print len(lowest_level_dict.keys())\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "63\n"
       ]
      }
     ],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import string\n",
      " \n",
      "try:\n",
      "    import xml.etree.cElementTree as ET  # Try to use the C implementation.\n",
      "except ImportError:\n",
      "    import xml.etree.ElementTree as ET   # If unavailable, use the pure Python\n",
      "                                         # implementation.  \n",
      " \n",
      "### First define some helper functions ###\n",
      " \n",
      "# gmt_to_dict() converts the tab-delimited GMT input file to a dictionary object.\n",
      "def gmt_to_dict(infile):\n",
      "    gmtdict = {}           # Start with an empty dictionary.\n",
      " \n",
      "    # Loop through each gene set in the file...\n",
      "    for line in infile:    # For each line (gene set)...\n",
      "        linelist = line.split(\"\\t\")  # Split line words using a tab separator.\n",
      "                                     # The result is a list object.\n",
      " \n",
      "        # The next line looks at the 0th element of linelist (pathway name).\n",
      "        # It filters out all non-printable characters. This is necessary in\n",
      "        # order to prevent a fatal error in the immediate future.\n",
      "        name = filter(lambda x: x in string.printable, linelist[0])\n",
      "        content = linelist[1:len(linelist)]  # Variable 'content' contains everything\n",
      "                                             # except for the pathway name. This\n",
      "                                             # means content[0] is the pathway description,\n",
      "                                             # which is what we'll use as a placeholder\n",
      "                                             # for the hierarchy tree location.\n",
      "        gmtdict[name] = content              # Key 'pathway name' now points to its\n",
      "                                             # content in gmtdict.\n",
      " \n",
      "    return gmtdict\n",
      " \n",
      "# write_dict() writes gmtdict to a tab-delimited file on disk.\n",
      "def write_dict(mydict, outfile):\n",
      "    for entry in mydict.items():  # For every dictionary item (which is a tuple)...\n",
      "        #...write the pathway name (entry 0, type string) and its content (entry 1,\n",
      "        # type list) as a new line to outfile, all separated by tabs.\n",
      "        outfile.write(\"\\t\".join([entry[0], \"\\t\".join(entry[1])]))\n",
      " \n",
      "# update_gmtdict() tries to put tree hierarchy information of a pathway found\n",
      "# in the source XML file into the right location in gmtdict.\n",
      "# Attribute 'node' can be seen as a line from the Reactome XML source.\n",
      "def update_gmtdict(node, hierarchy):\n",
      "    if node.tag == \"Pathway\":               # Ignore Reactions and BlackBoxEvents\n",
      "        # Get current node's displayName attribute. These match 1:1 with gmtdict keys.\n",
      "        s = node.get(\"displayName\").strip() # Also strip any leading and trailing\n",
      "                                            # white spaces.\n",
      "        # Filter out all non-printable characters.\n",
      "        s = filter(lambda x: x in string.printable, s)\n",
      " \n",
      "        # Key may not exist since the XML contains more nodes than there are\n",
      "        # pathways in the GMT:\n",
      "        try:\n",
      "            # Convert the integers in the hierarchy list to strings using the\n",
      "            # map(str, ...) function. Then join them together separated by dots.\n",
      "            # Finally update this pathway's hierarchy, which is the 0th element of\n",
      "            # the list pointed to by key s.\n",
      "            gmtdict[s][0] = \".\".join(map(str, hierarchy))\n",
      "        except KeyError:\n",
      "            print(\"Info: KeyError - \" + s + \", skipping...\")\n",
      " \n",
      "# walk_xml_node() receives a node from the XML tree and immediately starts\n",
      "# looping through all of its sub-nodes. Its sub-node number is added to the\n",
      "# current hierarchy list which is then used to update the current pathway's\n",
      "# tree location in gmtdict. Then, the function recursively walks all of the\n",
      "# current sub-node's sub-nodes, creating deeper and deeper hierarchies.\n",
      "# Each time this function is called, it receives the node we are investigating,\n",
      "# the current hierarchy list and default sub-node number of 0.\n",
      "def walk_xml_node(node, hierarchy, subnodenr=0):\n",
      "    for element in node:\n",
      "        subnodenr += 1              # For each element, increment subnodenr by 1.\n",
      "        hierarchy.append(subnodenr)          # Add sub-node nr to hierarchy list.\n",
      "        update_gmtdict(element, hierarchy)   # Update pathway info in gmtdict.\n",
      "        walk_xml_node(element, hierarchy)    # Walk sub-node's sub-nodes.\n",
      " \n",
      "    # When we're finished cycling through a set of sub-nodes and the hierarchy\n",
      "    # list contains at least one value, remove the last value from the list.\n",
      "    # This means we're moving up one level in the tree hierarchy.\n",
      "    if len(hierarchy) > 0:\n",
      "        hierarchy.pop()\n",
      " \n",
      "###################### End of helper functions #################################\n",
      "########### This is typically where the fun really starts ######################\n",
      " \n",
      "# gmtinfile points to the location of the unzipped Reactome GMT file on your local disk.\n",
      "gmtinfile = open(\"ReactomePathways.gmt\", \"r\")       # Open file for reading.\n",
      " \n",
      "# gmtoutfile points to the desired location of this script's result file. It will be\n",
      "# basically about the same as gmtinfile, but the description column will contain the\n",
      "# textual representation of each pathway's position in the tree.\n",
      "gmtoutfile = open(\"ReactomePathwaysTree.gmt\", \"w\")  # Open file for writing.\n",
      " \n",
      "# XMLfileloc is the location of the full Homo sapiens data set we obtained through the\n",
      "# RESTful API.\n",
      "XMLfileloc = \"REACTOME_tree.xml\"  # No file handle, just a string.\n",
      " \n",
      "gmtdict = gmt_to_dict(gmtinfile) # Put the GMT source file into a dictionary object.\n",
      "tree = ET.ElementTree(file=XMLfileloc)  # Load the XML tree as a ElementTree object.\n",
      "root = tree.getroot()        # Get the first node element (root) and start from there.\n",
      " \n",
      "# walk_xml_node() is a recursive function which walks through the ElementTree\n",
      "# representation of the XML tree, keeping track of its position in the tree\n",
      "# hierarchy, which is put in a list object.\n",
      "walk_xml_node(root, [])  # Start at the beginning (root) without a hierarchy (empty list).\n",
      "write_dict(gmtdict, gmtoutfile)    # Write gmtdict to a tab-delimited text file on disk.\n",
      " \n",
      "gmtinfile.close      # Close input file handle.\n",
      "gmtoutfile.close()   # Close output file handle."
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Info: KeyError - Loss of proteins required for interphase microtubule organization from the centrosome, skipping...\n",
        "Info: KeyError - PTEN Loss of Function in Cancer, skipping...\n",
        "Info: KeyError - Mucopolysaccharidoses, skipping...\n",
        "Info: KeyError - MPS I - Hurler syndrome, skipping...\n",
        "Info: KeyError - MPS II - Hunter syndrome, skipping...\n",
        "Info: KeyError - MPS IIIA - Sanfilippo syndrome A, skipping...\n",
        "Info: KeyError - MPS IIIB - Sanfilippo syndrome B, skipping...\n",
        "Info: KeyError - MPS IIIC - Sanfilippo syndrome C, skipping...\n",
        "Info: KeyError - MPS IIID - Sanfilippo syndrome D, skipping...\n",
        "Info: KeyError - MPS IV - Morquio syndrome A, skipping...\n",
        "Info: KeyError - MPS IV - Morquio syndrome B, skipping...\n",
        "Info: KeyError - MPS VI - Maroteaux-Lamy syndrome, skipping...\n",
        "Info: KeyError - MPS VII - Sly syndrome, skipping...\n",
        "Info: KeyError - MPS IX - Natowicz syndrome, skipping...\n",
        "Info: KeyError - Glycogen storage disease type II (GAA), skipping...\n",
        "Info: KeyError - Hereditary fructose intolerance, skipping...\n",
        "Info: KeyError - Essential fructosuria, skipping...\n",
        "Info: KeyError - Intestinal saccharidase deficiencies, skipping...\n",
        "Info: KeyError - Phenylketonuria, skipping...\n",
        "Info: KeyError - Defective LMBRD1 causes methylmalonic aciduria and homocystinuria type cblF, skipping...\n",
        "Info: KeyError - Defective MMAB causes methylmalonic aciduria type cblB, skipping...\n",
        "Info: KeyError - Defective BTD causes biotidinase deficiency, skipping...\n",
        "Info: KeyError - Defective ACY1 causes encephalopathy, skipping...\n",
        "Info: KeyError - Defective AHCY causes Hypermethioninemia with S-adenosylhomocysteine hydrolase deficiency (HMAHCHD), skipping...\n",
        "Info: KeyError - Defective CYP11B1 causes Adrenal hyperplasia 4 (AH4), skipping...\n",
        "Info: KeyError - Defective CYP11B2 causes Corticosterone methyloxidase 1 deficiency (CMO-1 deficiency), skipping...\n",
        "Info: KeyError - Defective CYP17A1 causes Adrenal hyperplasia 5 (AH5), skipping...\n",
        "Info: KeyError - Defective CYP19A1 causes Aromatase excess syndrome (AEXS), skipping...\n",
        "Info: KeyError - Defective CYP1B1 causes Glaucoma, skipping...\n",
        "Info: KeyError - Defective CYP21A2 causes Adrenal hyperplasia 3 (AH3), skipping...\n",
        "Info: KeyError - Defective CYP24A1 causes Hypercalcemia, infantile (HCAI), skipping...\n",
        "Info: KeyError - Defective CYP26B1 causes Radiohumeral fusions with other skeletal and craniofacial anomalies (RHFCA), skipping...\n",
        "Info: KeyError - Defective CYP26C1 causes Focal facial dermal dysplasia 4 (FFDD4), skipping...\n",
        "Info: KeyError - Defective CYP27A1 causes Cerebrotendinous xanthomatosis (CTX), skipping...\n",
        "Info: KeyError - Defective CYP27B1 causes Rickets vitamin D-dependent 1A (VDDR1A), skipping...\n",
        "Info: KeyError - Defective CYP2R1 causes Rickets vitamin D-dependent 1B (VDDR1B), skipping...\n",
        "Info: KeyError - Defective CYP2U1 causes Spastic paraplegia 56, autosomal recessive (SPG56), skipping...\n",
        "Info: KeyError - Defective CYP4F22 causes Ichthyosis, congenital, autosomal recessive 5 (ARCI5), skipping...\n",
        "Info: KeyError - Defective TBXAS1 causes Ghosal hematodiaphyseal dysplasia (GHDD), skipping...\n",
        "Info: KeyError - Defective CYP7B1 causes Spastic paraplegia 5A, autosomal recessive (SPG5A) and Congenital bile acid synthesis defect 3 (CBAS3), skipping...\n",
        "Info: KeyError - Defective FMO3 causes Trimethylaminuria (TMAU), skipping...\n",
        "Info: KeyError - Defective GCLC causes Hemolytic anemia due to gamma-glutamylcysteine synthetase deficiency (HAGGSD), skipping...\n",
        "Info: KeyError - Defective GGT1 causes Glutathionuria (GLUTH), skipping...\n",
        "Info: KeyError - Defective GSS causes Glutathione synthetase deficiency (GSS deficiency), skipping...\n",
        "Info: KeyError - Defective OPLAH causes 5-oxoprolinase deficiency (OPLAHD), skipping...\n",
        "Info: KeyError - Defective MAOA causes Brunner syndrome (BRUNS), skipping...\n",
        "Info: KeyError - Defective MAT1A causes Methionine adenosyltransferase deficiency (MATD), skipping...\n",
        "Info: KeyError - Defective SLC35D1 causes Schneckenbecken dysplasia (SCHBCKD), skipping...\n",
        "Info: KeyError - Defective TPMT causes Thiopurine S-methyltransferase deficiency (TPMT deficiency), skipping...\n",
        "Info: KeyError - Defective UGT1A1 causes hyperbilirubinemia, skipping...\n",
        "Info: KeyError - Defective UGT1A4 causes hyperbilirubinemia, skipping...\n",
        "Info: KeyError - Defective SLC26A2 causes chondrodysplasias, skipping...\n",
        "Info: KeyError - Defective PAPSS2 causes SEMD-PA, skipping...\n",
        "Info: KeyError - Diseases associated with N-glycosylation of proteins, skipping...\n",
        "Info: KeyError - Defective ALG6 causes ALG6-CDG (CDG-1c), skipping...\n",
        "Info: KeyError - Defective ALG3 causes ALG3-CDG (CDG-1d), skipping...\n",
        "Info: KeyError - Defective MPDU1 causes MPDU1-CDG (CDG-1f), skipping...\n",
        "Info: KeyError - Defective ALG12 causes ALG12-CDG (CDG-1g), skipping...\n",
        "Info: KeyError - Defective ALG8 causes ALG8-CDG (CDG-1h), skipping...\n",
        "Info: KeyError - Defective ALG2 causes ALG2-CDG (CDG-1i), skipping...\n",
        "Info: KeyError - Defective DPAGT1 causes DPAGT1-CDG (CDG-1j) and CMSTA2, skipping...\n",
        "Info: KeyError - Defective ALG1 causes ALG1-CDG (CDG-1k), skipping...\n",
        "Info: KeyError - Defective ALG9 causes ALG9-CDG (CDG-1l), skipping...\n",
        "Info: KeyError - Defective RFT1 causes RFT1-CDG (CDG-1n), skipping...\n",
        "Info: KeyError - Defective ALG11 causes ALG11-CDG (CDG-1p), skipping...\n",
        "Info: KeyError - Defective MGAT2 causes MGAT2-CDG (CDG-2a), skipping...\n",
        "Info: KeyError - Defective MOGS causes MOGS-CDG (CDG-2b), skipping...\n",
        "Info: KeyError - Defective MAN1B1 causes MRT15, skipping...\n",
        "Info: KeyError - Defective ALG14 causes congenital myasthenic syndrome (ALG14-CMS), skipping...\n",
        "Info: KeyError - HIV Transcription Termination, skipping...\n",
        "Info: KeyError - Resolution of D-loop structures through synthesis-dependent strand-annealing, skipping...\n",
        "Info: KeyError - IP3 and IP4 transport between cytosol and nucleus, skipping...\n",
        "Info: KeyError - IPs transport between nucleus and cytosol, skipping...\n",
        "Info: KeyError - IP6 and IP7 transport between cytosol and nucleus, skipping...\n",
        "Info: KeyError - IPs transport between cytosol and ER lumen, skipping...\n",
        "Info: KeyError - IPs transport between ER lumen and cytosol, skipping...\n",
        "Info: KeyError - IPs transport between ER lumen and nucleus, skipping...\n",
        "Info: KeyError - IPs transport between nucleus and ER lumen, skipping...\n",
        "Info: KeyError - Synthesis of BMP, skipping...\n",
        "Info: KeyError - PIPs transport between Golgi and plasma membranes, skipping...\n",
        "Info: KeyError - PIPs transport between plasma and early endosome membranes, skipping...\n",
        "Info: KeyError - PIPs transport between early endosome and Golgi membranes, skipping...\n",
        "Info: KeyError - PIPs transport between early and late endosome membranes, skipping...\n",
        "Info: KeyError - PIPs transport between late endosome and Golgi membranes, skipping...\n",
        "Info: KeyError - Feedback control in the MAPK signaling module, skipping...\n"
       ]
      }
     ],
     "prompt_number": 10
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_list=[]\n",
      "with open('ReactomePathwaysTree.txt') as f:\n",
      "    for line in f:\n",
      "        gene_list=gene_list+line.strip().split('\\t')[2:]\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "75493\n",
        "7855\n"
       ]
      }
     ],
     "prompt_number": 16
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from collections import Counter\n",
      "level_count=Counter()\n",
      "with open('ReactomePathwaysTree.txt') as f:\n",
      "    for line in f:\n",
      "        index = line.split('\\t')[1].split('.')\n",
      "        level_count[len(index)]+=len(line.split('\\t')[2:])\n",
      "print level_count"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Counter({3: 18473, 2: 15899, 4: 15843, 1: 13104, 5: 8051, 6: 2390, 7: 1163, 8: 389, 9: 111, 10: 65, 11: 5})\n"
       ]
      }
     ],
     "prompt_number": 25
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import numpy as np\n",
      "pathways=0\n",
      "gene_counts = []\n",
      "genes=[]\n",
      "with open('c2.cp.kegg.v5.0.entrez.gmt') as f:\n",
      "    for line in f:\n",
      "        pathways+=1\n",
      "        print line.strip().split('\\t')[0]\n",
      "        gene_counts.append(len(line.strip().split('\\t'))-2)\n",
      "        genes=genes+line.strip().split('\\t')[2:]\n",
      "print pathways\n",
      "print np.mean(np.array(gene_counts))\n",
      "print len(genes)\n",
      "print len(list(set(genes)))\n",
      "\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "KEGG_GLYCOLYSIS_GLUCONEOGENESIS\n",
        "KEGG_CITRATE_CYCLE_TCA_CYCLE\n",
        "KEGG_PENTOSE_PHOSPHATE_PATHWAY\n",
        "KEGG_PENTOSE_AND_GLUCURONATE_INTERCONVERSIONS\n",
        "KEGG_FRUCTOSE_AND_MANNOSE_METABOLISM\n",
        "KEGG_GALACTOSE_METABOLISM\n",
        "KEGG_ASCORBATE_AND_ALDARATE_METABOLISM\n",
        "KEGG_FATTY_ACID_METABOLISM\n",
        "KEGG_STEROID_BIOSYNTHESIS\n",
        "KEGG_PRIMARY_BILE_ACID_BIOSYNTHESIS\n",
        "KEGG_STEROID_HORMONE_BIOSYNTHESIS\n",
        "KEGG_OXIDATIVE_PHOSPHORYLATION\n",
        "KEGG_PURINE_METABOLISM\n",
        "KEGG_PYRIMIDINE_METABOLISM\n",
        "KEGG_ALANINE_ASPARTATE_AND_GLUTAMATE_METABOLISM\n",
        "KEGG_GLYCINE_SERINE_AND_THREONINE_METABOLISM\n",
        "KEGG_CYSTEINE_AND_METHIONINE_METABOLISM\n",
        "KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_DEGRADATION\n",
        "KEGG_VALINE_LEUCINE_AND_ISOLEUCINE_BIOSYNTHESIS\n",
        "KEGG_LYSINE_DEGRADATION\n",
        "KEGG_ARGININE_AND_PROLINE_METABOLISM\n",
        "KEGG_HISTIDINE_METABOLISM\n",
        "KEGG_TYROSINE_METABOLISM\n",
        "KEGG_PHENYLALANINE_METABOLISM\n",
        "KEGG_TRYPTOPHAN_METABOLISM\n",
        "KEGG_BETA_ALANINE_METABOLISM\n",
        "KEGG_TAURINE_AND_HYPOTAURINE_METABOLISM\n",
        "KEGG_SELENOAMINO_ACID_METABOLISM\n",
        "KEGG_GLUTATHIONE_METABOLISM\n",
        "KEGG_STARCH_AND_SUCROSE_METABOLISM\n",
        "KEGG_N_GLYCAN_BIOSYNTHESIS\n",
        "KEGG_OTHER_GLYCAN_DEGRADATION\n",
        "KEGG_O_GLYCAN_BIOSYNTHESIS\n",
        "KEGG_AMINO_SUGAR_AND_NUCLEOTIDE_SUGAR_METABOLISM\n",
        "KEGG_GLYCOSAMINOGLYCAN_DEGRADATION\n",
        "KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_CHONDROITIN_SULFATE\n",
        "KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_KERATAN_SULFATE\n",
        "KEGG_GLYCOSAMINOGLYCAN_BIOSYNTHESIS_HEPARAN_SULFATE\n",
        "KEGG_GLYCEROLIPID_METABOLISM\n",
        "KEGG_INOSITOL_PHOSPHATE_METABOLISM\n",
        "KEGG_GLYCOSYLPHOSPHATIDYLINOSITOL_GPI_ANCHOR_BIOSYNTHESIS\n",
        "KEGG_GLYCEROPHOSPHOLIPID_METABOLISM\n",
        "KEGG_ETHER_LIPID_METABOLISM\n",
        "KEGG_ARACHIDONIC_ACID_METABOLISM\n",
        "KEGG_LINOLEIC_ACID_METABOLISM\n",
        "KEGG_ALPHA_LINOLENIC_ACID_METABOLISM\n",
        "KEGG_SPHINGOLIPID_METABOLISM\n",
        "KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_LACTO_AND_NEOLACTO_SERIES\n",
        "KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_GLOBO_SERIES\n",
        "KEGG_GLYCOSPHINGOLIPID_BIOSYNTHESIS_GANGLIO_SERIES\n",
        "KEGG_PYRUVATE_METABOLISM\n",
        "KEGG_GLYOXYLATE_AND_DICARBOXYLATE_METABOLISM\n",
        "KEGG_PROPANOATE_METABOLISM\n",
        "KEGG_BUTANOATE_METABOLISM\n",
        "KEGG_ONE_CARBON_POOL_BY_FOLATE\n",
        "KEGG_RIBOFLAVIN_METABOLISM\n",
        "KEGG_NICOTINATE_AND_NICOTINAMIDE_METABOLISM\n",
        "KEGG_PANTOTHENATE_AND_COA_BIOSYNTHESIS\n",
        "KEGG_FOLATE_BIOSYNTHESIS\n",
        "KEGG_RETINOL_METABOLISM\n",
        "KEGG_PORPHYRIN_AND_CHLOROPHYLL_METABOLISM\n",
        "KEGG_TERPENOID_BACKBONE_BIOSYNTHESIS\n",
        "KEGG_LIMONENE_AND_PINENE_DEGRADATION\n",
        "KEGG_NITROGEN_METABOLISM\n",
        "KEGG_SULFUR_METABOLISM\n",
        "KEGG_AMINOACYL_TRNA_BIOSYNTHESIS\n",
        "KEGG_METABOLISM_OF_XENOBIOTICS_BY_CYTOCHROME_P450\n",
        "KEGG_DRUG_METABOLISM_CYTOCHROME_P450\n",
        "KEGG_DRUG_METABOLISM_OTHER_ENZYMES\n",
        "KEGG_BIOSYNTHESIS_OF_UNSATURATED_FATTY_ACIDS\n",
        "KEGG_ABC_TRANSPORTERS\n",
        "KEGG_RIBOSOME\n",
        "KEGG_RNA_DEGRADATION\n",
        "KEGG_RNA_POLYMERASE\n",
        "KEGG_BASAL_TRANSCRIPTION_FACTORS\n",
        "KEGG_DNA_REPLICATION\n",
        "KEGG_SPLICEOSOME\n",
        "KEGG_PROTEASOME\n",
        "KEGG_PROTEIN_EXPORT\n",
        "KEGG_PPAR_SIGNALING_PATHWAY\n",
        "KEGG_BASE_EXCISION_REPAIR\n",
        "KEGG_NUCLEOTIDE_EXCISION_REPAIR\n",
        "KEGG_MISMATCH_REPAIR\n",
        "KEGG_HOMOLOGOUS_RECOMBINATION\n",
        "KEGG_NON_HOMOLOGOUS_END_JOINING\n",
        "KEGG_MAPK_SIGNALING_PATHWAY\n",
        "KEGG_ERBB_SIGNALING_PATHWAY\n",
        "KEGG_CALCIUM_SIGNALING_PATHWAY\n",
        "KEGG_CYTOKINE_CYTOKINE_RECEPTOR_INTERACTION\n",
        "KEGG_CHEMOKINE_SIGNALING_PATHWAY\n",
        "KEGG_PHOSPHATIDYLINOSITOL_SIGNALING_SYSTEM\n",
        "KEGG_NEUROACTIVE_LIGAND_RECEPTOR_INTERACTION\n",
        "KEGG_CELL_CYCLE\n",
        "KEGG_OOCYTE_MEIOSIS\n",
        "KEGG_P53_SIGNALING_PATHWAY\n",
        "KEGG_UBIQUITIN_MEDIATED_PROTEOLYSIS\n",
        "KEGG_SNARE_INTERACTIONS_IN_VESICULAR_TRANSPORT\n",
        "KEGG_REGULATION_OF_AUTOPHAGY\n",
        "KEGG_LYSOSOME\n",
        "KEGG_ENDOCYTOSIS\n",
        "KEGG_PEROXISOME\n",
        "KEGG_MTOR_SIGNALING_PATHWAY\n",
        "KEGG_APOPTOSIS\n",
        "KEGG_CARDIAC_MUSCLE_CONTRACTION\n",
        "KEGG_VASCULAR_SMOOTH_MUSCLE_CONTRACTION\n",
        "KEGG_WNT_SIGNALING_PATHWAY\n",
        "KEGG_DORSO_VENTRAL_AXIS_FORMATION\n",
        "KEGG_NOTCH_SIGNALING_PATHWAY\n",
        "KEGG_HEDGEHOG_SIGNALING_PATHWAY\n",
        "KEGG_TGF_BETA_SIGNALING_PATHWAY\n",
        "KEGG_AXON_GUIDANCE\n",
        "KEGG_VEGF_SIGNALING_PATHWAY\n",
        "KEGG_FOCAL_ADHESION\n",
        "KEGG_ECM_RECEPTOR_INTERACTION\n",
        "KEGG_CELL_ADHESION_MOLECULES_CAMS\n",
        "KEGG_ADHERENS_JUNCTION\n",
        "KEGG_TIGHT_JUNCTION\n",
        "KEGG_GAP_JUNCTION\n",
        "KEGG_COMPLEMENT_AND_COAGULATION_CASCADES\n",
        "KEGG_ANTIGEN_PROCESSING_AND_PRESENTATION\n",
        "KEGG_RENIN_ANGIOTENSIN_SYSTEM\n",
        "KEGG_TOLL_LIKE_RECEPTOR_SIGNALING_PATHWAY\n",
        "KEGG_NOD_LIKE_RECEPTOR_SIGNALING_PATHWAY\n",
        "KEGG_RIG_I_LIKE_RECEPTOR_SIGNALING_PATHWAY\n",
        "KEGG_CYTOSOLIC_DNA_SENSING_PATHWAY\n",
        "KEGG_JAK_STAT_SIGNALING_PATHWAY\n",
        "KEGG_HEMATOPOIETIC_CELL_LINEAGE\n",
        "KEGG_NATURAL_KILLER_CELL_MEDIATED_CYTOTOXICITY\n",
        "KEGG_T_CELL_RECEPTOR_SIGNALING_PATHWAY\n",
        "KEGG_B_CELL_RECEPTOR_SIGNALING_PATHWAY\n",
        "KEGG_FC_EPSILON_RI_SIGNALING_PATHWAY\n",
        "KEGG_FC_GAMMA_R_MEDIATED_PHAGOCYTOSIS\n",
        "KEGG_LEUKOCYTE_TRANSENDOTHELIAL_MIGRATION\n",
        "KEGG_INTESTINAL_IMMUNE_NETWORK_FOR_IGA_PRODUCTION\n",
        "KEGG_CIRCADIAN_RHYTHM_MAMMAL\n",
        "KEGG_LONG_TERM_POTENTIATION\n",
        "KEGG_NEUROTROPHIN_SIGNALING_PATHWAY\n",
        "KEGG_LONG_TERM_DEPRESSION\n",
        "KEGG_OLFACTORY_TRANSDUCTION\n",
        "KEGG_TASTE_TRANSDUCTION\n",
        "KEGG_REGULATION_OF_ACTIN_CYTOSKELETON\n",
        "KEGG_INSULIN_SIGNALING_PATHWAY\n",
        "KEGG_GNRH_SIGNALING_PATHWAY\n",
        "KEGG_PROGESTERONE_MEDIATED_OOCYTE_MATURATION\n",
        "KEGG_MELANOGENESIS\n",
        "KEGG_ADIPOCYTOKINE_SIGNALING_PATHWAY\n",
        "KEGG_TYPE_II_DIABETES_MELLITUS\n",
        "KEGG_TYPE_I_DIABETES_MELLITUS\n",
        "KEGG_MATURITY_ONSET_DIABETES_OF_THE_YOUNG\n",
        "KEGG_ALDOSTERONE_REGULATED_SODIUM_REABSORPTION\n",
        "KEGG_VASOPRESSIN_REGULATED_WATER_REABSORPTION\n",
        "KEGG_PROXIMAL_TUBULE_BICARBONATE_RECLAMATION\n",
        "KEGG_ALZHEIMERS_DISEASE\n",
        "KEGG_PARKINSONS_DISEASE\n",
        "KEGG_AMYOTROPHIC_LATERAL_SCLEROSIS_ALS\n",
        "KEGG_HUNTINGTONS_DISEASE\n",
        "KEGG_PRION_DISEASES\n",
        "KEGG_VIBRIO_CHOLERAE_INFECTION\n",
        "KEGG_EPITHELIAL_CELL_SIGNALING_IN_HELICOBACTER_PYLORI_INFECTION\n",
        "KEGG_PATHOGENIC_ESCHERICHIA_COLI_INFECTION\n",
        "KEGG_LEISHMANIA_INFECTION\n",
        "KEGG_PATHWAYS_IN_CANCER\n",
        "KEGG_COLORECTAL_CANCER\n",
        "KEGG_RENAL_CELL_CARCINOMA\n",
        "KEGG_PANCREATIC_CANCER\n",
        "KEGG_ENDOMETRIAL_CANCER\n",
        "KEGG_GLIOMA\n",
        "KEGG_PROSTATE_CANCER\n",
        "KEGG_THYROID_CANCER\n",
        "KEGG_BASAL_CELL_CARCINOMA\n",
        "KEGG_MELANOMA\n",
        "KEGG_BLADDER_CANCER\n",
        "KEGG_CHRONIC_MYELOID_LEUKEMIA\n",
        "KEGG_ACUTE_MYELOID_LEUKEMIA\n",
        "KEGG_SMALL_CELL_LUNG_CANCER\n",
        "KEGG_NON_SMALL_CELL_LUNG_CANCER\n",
        "KEGG_ASTHMA\n",
        "KEGG_AUTOIMMUNE_THYROID_DISEASE\n",
        "KEGG_SYSTEMIC_LUPUS_ERYTHEMATOSUS\n",
        "KEGG_ALLOGRAFT_REJECTION\n",
        "KEGG_GRAFT_VERSUS_HOST_DISEASE\n",
        "KEGG_PRIMARY_IMMUNODEFICIENCY\n",
        "KEGG_HYPERTROPHIC_CARDIOMYOPATHY_HCM\n",
        "KEGG_ARRHYTHMOGENIC_RIGHT_VENTRICULAR_CARDIOMYOPATHY_ARVC\n",
        "KEGG_DILATED_CARDIOMYOPATHY\n",
        "KEGG_VIRAL_MYOCARDITIS\n",
        "186\n",
        "69.2204301075\n",
        "12875\n",
        "5267\n"
       ]
      }
     ],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "new_kegg=[]\n",
      "with open('kegg_api.txt') as f:\n",
      "    for line in f:\n",
      "        new_kegg.append(line.split('\\t')[1])\n",
      "\n",
      "print len(new_kegg)\n",
      "print new_kegg[:10]\n",
      "print len(set(new_kegg))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "24825\n",
        "['hsa:10327\\n', 'hsa:124\\n', 'hsa:125\\n', 'hsa:126\\n', 'hsa:127\\n', 'hsa:128\\n', 'hsa:130\\n', 'hsa:130589\\n', 'hsa:131\\n', 'hsa:160287\\n']\n",
        "6898\n"
       ]
      }
     ],
     "prompt_number": 13
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}